File: backend/app/schemas/audit.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel


class AuditUserBrief(BaseModel):
    id: int
    username: str


class AuditLogEntry(BaseModel):
    id: int
    user: AuditUserBrief | None
    action: str
    entity_type: str | None
    entity_id: int | None
    metadata: dict[str, Any] | None
    ip_address: str | None
    user_agent: str | None
    created_at: datetime


class AuditLogList(BaseModel):
    items: list[AuditLogEntry]
    total: int
    page: int
    page_size: int

File: frontend/src/api/audit.ts
import { apiClient } from './client';

export interface AuditUser {
  id: number;
  username: string;
}

export interface AuditEntry {
  id: number;
  user: AuditUser | null;
  action: string;
  entity_type: string | null;
  entity_id: number | null;
  metadata: Record<string, unknown> | null;
  ip_address: string | null;
  user_agent: string | null;
  created_at: string;
}

export interface AuditListResponse {
  items: AuditEntry[];
  total: number;
  page: number;
  page_size: number;
}

export async function listAuditLog(params: {
  page: number;
  page_size: number;
  action?: string;
  entity_type?: string;
}): Promise<AuditListResponse> {
  const { data } = await apiClient.get<AuditListResponse>('/admin/audit-log', { params });
  return data;
}
